File: 7web/code/sinatra/validation/app_test.rb
#---
# Excerpted from "Seven Web Frameworks in Seven Weeks",
# published by The Pragmatic Bookshelf.
# Copyrights apply to this code. It may not be used to create training material, 
# courses, books, articles, and the like. Contact us if you are in doubt.
# We make no guarantees that this code is fit for any purpose. 
# Visit http://www.pragmaticprogrammer.com/titles/7web for more book information.
#---
require_relative "app"
require "rspec"
require "rack/test"
require "json"

describe "Bookmarking App" do
  include Rack::Test::Methods

  def app
    Sinatra::Application
  end

  it "returns a list of bookmarks" do
    get "/bookmarks"
    last_response.should be_ok
    bookmarks = JSON.parse(last_response.body)
    bookmarks.should be_instance_of(Array)
  end

  it "creates a new bookmark" do
    get "/bookmarks"
    bookmarks = JSON.parse(last_response.body)
    last_size = bookmarks.size

    post "/bookmarks",
      {:url => "http://www.test.com", :title => "Test"}
    last_response.status.should == 201
    last_response.body.should match(/\/bookmarks\/\d+/)

    get "/bookmarks"
    bookmarks = JSON.parse(last_response.body)
    expect(bookmarks.size).to eq(last_size + 1)
  end

  it "updates a bookmark" do
    url = "http://www.test.com"
    post "/bookmarks", {:url => url, :title => "Test"}
    bookmark_uri = last_response.body
    id = bookmark_uri.split("/").last
    
    put "/bookmarks/#{id}", {:url => url, :title => "Success"}
    last_response.status.should == 204

    get "/bookmarks/#{id}"
    retrieved_bookmark = JSON.parse(last_response.body)
    expect(retrieved_bookmark["title"]).to eq("Success")
  end

  it "deletes a bookmark" do
    post "/bookmarks",
      {:url => "http://www.test.com", :title => "Test"}
    get "/bookmarks"
    bookmarks = JSON.parse(last_response.body)
    last_size = bookmarks.size
    
    delete "/bookmarks/#{bookmarks.last['id']}"
    last_response.status.should == 200

    get "/bookmarks"
    bookmarks = JSON.parse(last_response.body)
    expect(bookmarks.size).to eq(last_size - 1)
  end

  it "sends an error code for an invalid get request" do
    get "/bookmarks/0"
    last_response.status.should == 404
  end

  it "sends an error code for an invalid put request" do
    put "/bookmarks/0", {:title => "Success"}
    last_response.status.should == 404
  end

  it "sends an error code for an invalid delete request" do
    delete "/bookmarks/0"
    last_response.status.should == 404
  end

  it "sends an error code for an invalid create request" do
    post "/bookmarks", {:url => "test", :title => "Test"}
    last_response.status.should == 400
  end

  it "sends an error code for an invalid update request" do
    get "/bookmarks"
    bookmarks = JSON.parse(last_response.body)
    id = bookmarks.first['id']

    put "/bookmarks/#{id}", {:url => "Invalid"}
    last_response.status.should == 400
  end
end

File: 7web/code/canjs/public/app/tagfilter/app-test.js
/***
 * Excerpted from "Seven Web Frameworks in Seven Weeks",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material, 
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose. 
 * Visit http://www.pragmaticprogrammer.com/titles/7web for more book information.
***/
describe("tagfilter/app-test.js", function() {
  describe("Bookmark", function() {
    describe("tags", function() {
      var bookmark = null;
      var tagList = null;

      beforeEach(function() {
        tagList = new can.Observe.List(["One", "Two"]);
        bookmark = new TaggedBookmark({tagList:tagList});
      });
       
      it("initializes tagsAsString from tagList", function() {
        expect(bookmark.attr("tagsAsString")).toBe("One, Two");
      });

      it("sets ordered tagList from tagsAsString change", function() {
        bookmark.attr("tagsAsString", "Three, Four");
        expect(tagList.length).toBe(2);
        expect(tagList[0], "Four");
        expect(tagList[1], "Three");
      });

      it("filters out blanks from tagsAsString", function() {
        bookmark.attr("tagsAsString", ", Four, , ,,");
        expect(tagList.length).toBe(1);
        expect(tagList[0], "Four");
      });

      it("sets empty tagList from blank or just commas", function() {
        bookmark.attr("tagsAsString", ", , ");
        expect(tagList.length).toBe(0);

        bookmark.attr("tagsAsString", " ");
        expect(tagList.length).toBe(0);
      });
    });
  });
  describe("app/tagfilter/bookmark_list view template", function() {
    it("displays a list of bookmarks with link tags", function() {
      var tagList = ["Tag 1", "Tag 2"];
      var bookmarks = [
        {id: 1, url:"http://one.com", title:"One", tagList:tagList},
        {id: 2, url:"http://two.com", title:"Two"}
      ];
      var model = {bookmarks:bookmarks};
      var el = $("<div/>").append("/app/tagfilter/bookmark_list", model);

      expect(el.find("li").size()).toBe(bookmarks.length);
      expect(el.find("a.tag").size()).toBe(tagList.length);
      expect(el.find("a.tag:first").html()).toBe(tagList[0]);
    });
  });
  describe("TagFilterBookmarkListControl", function() {
    var eventHub = null;
    var bookmark = null;
    var element = null;
    var control = null;
    var filterObject = null;

    beforeEach(function() {
      eventHub = new can.Observe({});
      element = $("<div/>");
      bookmark = new Bookmark({url:"http://one.com", title:"One",
        tagList:["Tag A", "Tag B"]});
      var bookmarks = new Bookmark.List([bookmark]);
      filterObject = new can.Observe({});
      var options = {eventHub:eventHub, bookmarks:bookmarks,
        filterObject:filterObject};
      control = new TagFilterBookmarkListControl(element, options);

      eventHub.attr("bookmarks", bookmarks);
    });

    it("triggers filterTag", function(done) {
      filterObject.bind("filterTag", function() { done(); });
      element.find("a.tag").click();
    });
  });
  describe("tagfilter/tag_list view template", function() {
    it("displays a list of tags with bookmark counts and links", function() {
      var selector = "li:first a";
      var tags = [
        {label:"One", bookmarkCount:1},
        {label:"Two", bookmarkCount:2}
      ];
      var model = {bookmarks:{tags:tags}};
      var el = $("<div/>").append("/app/tagfilter/tag_list", model);
      expect(el.find("li").size()).toBe(tags.length);
      expect(el.find(selector).html()).toBe("One (1)");
    });
  });
  describe("TagListControl", function() {
    it("renders the tag list", function() {
      var el = $("<div/>");
      var bookmarks = new TaggedBookmark.List([
        new TaggedBookmark({url:"http://one.com", title:"One",
          tagList:new can.Observe.List(["Tag2","Tag1"])}),
        new TaggedBookmark({url:"http://two.com", title:"Two",
          tagList:new can.Observe.List(["Tag3","Tag1"])})]);
      new TagListControl(el, {bookmarks:bookmarks});
      expect(el.find("li:first a").html()).toBe("Tag1 (2)");
    });
  });
  describe("App_tagfilter", function() {
    it("initializes (sanity check)", function() {
      new App_tagfilter();
      expect(true).toBe(true);
    });
  });
});
